### ISSUE TITLE
Do not forbid rpm to handle %posttrans scriptlets

### ISSUE THREAD

**Issue by mikhirev** [openSUSE/libzypp#101]
Currently libzypp collects `%posttrans` scriptlets when installing packages and handles them by itself instead letting rpm to run them in the end of transaction as desired. There is at least one case when this breaks correct upgrade procedure.

Consider upgrading rpm itself, when the new version is linked to new db version, which breaks backwards compatibility. To deal with this, `%posttrans` scriptlet removing old rpm database is added to the package. So old version of `rpm` installs all new packages including rpm, executes `%posttrans` and terminate correctly. Next time the new `rpm` is run, which re-creates the database using new stroage format and then operates correctly.

When using `zypper` for such upgrade, old version of `rpm` is run to upgrade a single package breaking compatibility, but the `%posttrans` scriptlet execution is delayed. Then new `rpm` is executed to upgrade next package, but it fails because it's unable to open the database (in old format). The upgrade procedure aborts.

The obvious solution is to run `%posttrans` scriptlet when it is intended to run — after completion of rpm transaction.

**Comment by mlandres** [comment]
Yes, we probably have to exclude rpm.

**Comment by mlandres** [comment]
OTOH, after discussing it with our rpm maintainer, `rpm` should not delay a database conversion because this would break packages calling `rpm -q` from within their '%pre/post' scripts (using the new `rpm`). So a database conversion would already take place in the `%postin`.

But both scenarios are weak. We tend to delay a decision until we're actually in need to migrate a rpmbd.

**Comment by mikhirev** [comment]
Well, first there's no database conversion as such, it is simply being removed. Second, this cannot be done in `%pre` or `%post` because rpm still uses the database at that time (I don't remember details, but I tried this).

As I understand, in SUSE rpm is always built with the same db version (very old), so there's no such problem. But for our purposes, we don't want to maintain an ancient db version for rpm. So we used the `%posttrans` scriptlet the same way as it is done in Fedora.

**Comment by mlandres** [comment]
I currently don't want to change the default behavior, but I could offer to respond to e.g. an environment variable like `ZYPP_IMMEDIATE_RPM_POSTTRANS`.  If it's set, we would not delay the `rpm` packages `%posttrans`. Would that help you?

**Comment by mikhirev** [comment]
In our build of libzypp, we simply reverted the commit that introduced deferring `%posttrans`. That's a better solution for us, but the purpose of reporting this issue report is to reduce differences between our fork and upstream, and to prevent someone else from running into the same issue. So if you don't want to change default libzypp behavior, we don't need any additional tricks like "magic" environment variables, we simply continue using or fork. But this does not seem a right solution for me.

BTW I dont' understand the purpose of deferring `%posttrans`. I found old tickets in bugzilla that reported slow operation of `%posttrans` when it is called multiple times, but that was caused by incorrect packaging: things that slould be handled by triggers, were added to `%posttrans` instead.

**Comment by mlandres** [comment]
We don't use a librpm-transaction to install. Our solver builds the transaction and we call `rpm` repeatedly just to install/remove single packages.  That's why we must defer the `%posttrans`. Otherwise `rpm` will execute them immediately after each package got installed.

**Comment by mikhirev** [comment]
> We don't use a librpm-transaction to install.

You don't, but `rpm` does. ☺

> That's why we must defer the `%posttrans`. Otherwise `rpm` will execute them immediately after each package got installed.

And what's problem with that?

**Comment by mlandres** [comment]
Packages move actions into the `%posttrans` to be sure they get executed after all packages got installed and the system is in a consistent state again. If we'd let `rpm` execute the `%posttrans`, it would be would just be a 2nd kind of `%post`. Executed in the middle of the transaction where you very much depend on the install order.

**Comment by mikhirev** [comment]
Isn't the system in consistent state after each `rpm` command completion? It should be consistent at any moment unless there are circular dependencies between packages.

In fact, the main use case of `%posttrans` is executing script from new package *after* `%postun` from old package (because new `%post` is executed *before* old `%postun` and unable e. g. to fix some error in `%postun`). You are speaking about some other use case, can you explain it?

**Comment by mlschroe** [comment]
One of the main usages of posttrans is to reduce the number of times some script needs to be run, e.g. some index generation. That's why the gnome folks begged us to implement them like rpm does when some big transaction is used.

Using posttrans just to get something to run after the uninstall scriptlets is really just a hack.

**Comment by mikhirev** [comment]
> One of the main usages of posttrans is to reduce the number of times some script needs to be run, e.g. some index generation.

That's wrong. If several packages installed in single transaction has `%posttrans` scriptlet with the same command, each of these scripts will be executed, and the command will be run as many times, as many packages have it in `%posttrans`. In some cases the command can be clever enough to see that it has nothing to do when being run second and subsequent times, but anyway using `%posttrans` for this is bad practice. Such commands should be run from trigger scriptlets instead.

**Comment by mlandres** [comment]
> Isn't the system in consistent state after each rpm command completion?
If `rpm` steers a transaction, yes. If `libzypp` steers the transaction, it's in consistent state after the libzypp commit has ended. I.e. after all packages have been installed. 

> It should be consistent at any moment unless there are circular dependencies 
Not if you update systems. If a package requires a library, the new package may not work with the old library, and the old package not with the new lib. After updating the first part, the system is inconsistent until all parts are updated. Whether this inconsistency is harmful depends on whether the package is required in the meantime.

Basically we're executing the scripts in the same order as one big `rpm` transaction would do. The difference is that the `rpm` version installed on disk also unpacks the packages. So we expect `rpm` 
to work after it's installed on disk, and not just after the transaction has ended.

Not delaying the `rpm  %posttrans` could be a way to solve the issue, not delaying any `%posttrans` of course as well.

**Comment by Conan-Kudo** [comment]
This is kind of a problem for Fedora libzypp, too. Unlike SUSE rpm, we don't bundle and link a private rpmdb bdb, so you're basically asking for pain when this happens...
